pydantic/core_utils.py

```diff
@@ -0,0 +1,177 @@
+"""
+The goal is to move this file's contents into pydantic_core proper.
+"""
+
+from __future__ import annotations
+
+from typing import Callable
+
+from pydantic_core import CoreSchema, CoreSchemaType, core_schema
+from typing_extensions import get_args
+
+
+def apply_ref_substitutions(
+    schema: core_schema.CoreSchema, substitutions: dict[str, core_schema.CoreSchema]
+) -> core_schema.CoreSchema:
+    def _apply_substitution(s: core_schema.CoreSchema) -> core_schema.CoreSchema:
+        if s['type'] == 'recursive-ref':
+            return s
+        if 'ref' in s:
+            return substitutions.get(s['ref'], s)
+        return s
+
+    return WalkAndApply(_apply_substitution).walk(schema)
+
+
+def remove_duplicate_refs(schema: core_schema.CoreSchema) -> core_schema.CoreSchema:
+    """
+    Assumption: Any time that two schemas have the same ref, they are equivalent.
+    This function walks a schema recursively, replacing all but the first occurrence of each ref with
+    a recursive-ref schema referencing that ref.
+    """
+    visited: set[str] = set()
+
+    def _remove_duplicate_refs(s: core_schema.CoreSchema) -> core_schema.CoreSchema:
+        if s['type'] == 'recursive-ref':
+            return s
+        if 'ref' in s:
+            if s['ref'] in visited:
+                return {'schema_ref': s['ref'], 'type': 'recursive-ref'}
+            visited.add(s['ref'])
+        return s
+
+    return WalkAndApply(_remove_duplicate_refs).walk(schema)
+
+
+class WalkAndApply:
+    def __init__(
+        self, f: Callable[[core_schema.CoreSchema], core_schema.CoreSchema], apply_before_recurse: bool = True
+    ):
+        self.f = f
+
+        self.apply_before_recurse = apply_before_recurse
+
+        self._schema_type_to_method = self._build_schema_type_to_method()
+
+    def _build_schema_type_to_method(self) -> dict[CoreSchemaType, Callable[[CoreSchema], CoreSchema]]:
+        mapping: dict[CoreSchemaType, Callable[[CoreSchema], CoreSchema]] = {}
+        for key in get_args(CoreSchemaType):
+            method_name = f"handle_{key.replace('-', '_')}_schema"
+            mapping[key] = getattr(self, method_name, self._handle_other_schemas)
+        return mapping
+
+    def walk(self, schema: core_schema.CoreSchema) -> core_schema.CoreSchema:
+        return self._walk(schema)
+
+    def _walk(self, schema: core_schema.CoreSchema) -> core_schema.CoreSchema:
+        schema = schema.copy()
+        if self.apply_before_recurse:
+            schema = self.f(schema)
+        method = self._schema_type_to_method[schema['type']]
+        schema = method(schema)
+        if not self.apply_before_recurse:
+            schema = self.f(schema)
+        return schema
+
+    def _handle_other_schemas(self, schema: core_schema.CoreSchema) -> core_schema.CoreSchema:
```

Comment:
If we want to keep this file's code, I can replace this `_handle_other_schemas` method with the various type-specific methods it is currently covering, and we can add a test that ensures all schemas have methods defined (so adding new ones in the future would require us to update this class appropriately); I did it like this to reduce the amount of boilerplate I needed to test this approach out.

Reply:
TODO: Make sure new types get handled properly